### PR TITLE
Fix most mypy errors.

### PR DESCRIPTION
**Pull request by tristanlatr**

### CODE REVIEW

**Comment by tristanlatr** [comment]
Opinions on the two issues @adiroiban @mthuurne ?
```
pydoctor/epydoc/markup/restructuredtext.py:55: error: Module "docutils.parsers.rst" has no attribute "directives"; maybe "Directive"?  [attr-defined]
pydoctor/epydoc/markup/restructuredtext.py:395: error: Argument 1 to "OptionParser" has incompatible type "List[Any]"; expected "Optional[str]"  [arg-type]
```

**Comment by mthuurne** [comment]
Sorry for my slow responses; it's been a busy week and we had a bit of a heatwave here. I'll have a look now.

**Comment by mthuurne** [comment]
> ```
> pydoctor/epydoc/markup/restructuredtext.py:55: error: Module "docutils.parsers.rst" has no attribute "directives"; maybe "Directive"?  [attr-defined]
> pydoctor/epydoc/markup/restructuredtext.py:395: error: Argument 1 to "OptionParser" has incompatible type "List[Any]"; expected "Optional[str]"  [arg-type]
> ```

The first error is weird: I can reproduce it under `tox`, but in my manually maintained venv it doesn't occur, even if I'm using the same version of `mypy`, `docutils`, and `types-docutils`.

The second error I can reproduce in both setups.

**Comment by tristanlatr** [comment]
Ok. So let’s ignore them ?

**Comment by mthuurne** [comment]
> Ok. So let’s ignore them ?

I want to dig a bit deeper first.

**Comment by mthuurne** [comment]
The problem with `OptionParser` seems to be that `docutils.frontend.OptionParser` inherits from `optparse.OptionParser`, but has a different `__init__()` signature, which is missing from the type stubs. Therefore this one can be ignored indeed, with a link to https://github.com/python/typeshed/issues/5667 so we know what is happening and when we can remove the suppression.

**Comment by mthuurne** [comment]
`docutils.parsers.rst.directives` is a module, which is present in `docutils` itself but missing from `typing-docutils`. So this is another case of the `docutils` stubs being incomplete.

I still don't understand why I don't get the error in my dev venv: I thought the reason might be because I have `docutils` installed, but the `mypy` tox environment also installs `docutils` since it's a dependency of `sphinx`, which we depend on since they ship type annotations in their main distribution itself.

So ignoring the error would be the correct short-term solution here as well. You could link to https://github.com/python/typeshed/issues/1269 as documentation of the problem.

File: pydoctor/model.py
Comment (mthuurne) [inline pydoctor/model.py:811]: We should only use `assert` if we can we guarantee that the condition is always true. If we're not that sure, it would be better to raise an exception instead.

I think there is at least one unlikely but possible scenario in which this condition is not true: if a file is deleted or moved between the moment its name is discovered and the moment it is introspected.

File: pydoctor/templatewriter/__init__.py
Comment (mthuurne) [inline pydoctor/templatewriter/__init__.py:30]: I agree with the cast as a temporary solution, but we should also submit a PR to `typeshed` to add a return type for `parseString()`.

**Review by mthuurne — CHANGES_REQUESTED** [review]

File: pydoctor/model.py
Comment (tristanlatr) [reply pydoctor/model.py:811]: A RuntimeError is OK ?

**Comment by tristanlatr** [comment]
> You could link to python/typeshed#1269 as documentation of the problem.

I've tried using the [docutils-stubs](https://pypi.org/project/docutils-stubs/) as mentioned in this ticket but it's becoming a real headache, more than 50 errors that require a lot of additional checks in the restructuredtext module. I don't want to fix all mypy errors honestly. I will revert back to use `types-docutils` for the principal `mypy` env and create a separate one that uses `docutils-stubs` so we can look at it later. 

Also the HTMLTranslator class has been moved to pydoctor.node2stan._PydoctorHTMLTranslator in #386

File: tox.ini
Comment (glyph) [inline tox.ini:126]: Do we want an issue for this repo as well, to track turning it on in CI when this is done?

File: pydoctor/model.py
Comment (glyph) [inline pydoctor/model.py:812]: Can you do a test for this apparently untested case in a follow-up?

**Review by glyph — APPROVED** [review]
🚢

File: pydoctor/model.py
Comment (tristanlatr) [reply pydoctor/model.py:812]: Like @mthuurne was saying, this error is raised when  a file is deleted or moved between the moment its name is discovered and the moment it is introspected. So honestly I don't know how to test such a case.

File: tox.ini
Comment (tristanlatr) [reply tox.ini:126]: See https://github.com/twisted/pydoctor/issues/409

**Comment by tristanlatr** [comment]
Thank @glyph for your review. 

I'll let @mthuurne merge this PR. 

@glyph, If you want to help out reviewing more PRs, that would be great.  
There is #386 waiting for a review  ;)

**Comment by glyph** [comment]
> @glyph, If you want to help out reviewing more PRs, that would be great.

I'd love to, but WYSIWYG in this case — I already have a recurring task in my personal task management system to review open source PRs as frequently as I can, and … this is about it ;)

**Comment by tristanlatr** [comment]
@mthuurne Can I merge that ? Thanks

**Comment by tristanlatr** [comment]
@mthuurne I will merge this PR now since it's blocking the CI from being green on all other branches. 
If there is anything you'd like to be modified, please speak up and I'll submit a patch.

File: pydoctor/model.py
Comment (glyph) [reply pydoctor/model.py:812]: This is why `FilePath` narrows down its interface to a single object which returns other objects, so you can pass a parameter along to the system under test. In the absence of such compositional design, the way to test this is to mock out `spec_from_file_location` with `patch` in a test.

**Pull request merged**
